### ISSUE TITLE
Any plans to add custom loss functions?

### ISSUE THREAD

**Issue by Simplex-AP** [dmlc/MXNet.jl#77]
The python version of MXNet has the capability to define custom loss functions, but this is missing from the Julia version. Are there any plans to add it?

**Comment by vchuravy** [comment]
Depends on #39 and see my work in #16. The ability to define custom loss functions requires layers written in python and is what inspired me to work on it. I have plans to work on this but it will need to wait till julia v0.5

**Comment by pluskid** [comment]
Another option which might be easier to implement (though involves quite a lot refactoring but not technically difficult) is the new module system in the mxnet python side. See dmlc/mxnet#1868 for example. With the new module system, a hybrid symbolic and imperative module can be used. The computation graph is built using symbolic nodes, while the loss function is written directly in Python.

**Comment by Arkoniak** [comment]
There is a [discussion](https://github.com/dmlc/mxnet/issues/3368) In R branch of MXNet. They propose to use MakeLoss function. I've tried to implement their solution:
* [linear regression](https://gist.github.com/Arkoniak/5b9b33f399daae726bfe581cbd37fee1)
* [logistic regression](https://gist.github.com/Arkoniak/4ae1910dcb210e2b0db7fbd7027261ba)

Well, generally speaking it works, but it looks ugly. it would be nice, if anyone with deep knowledge of MXNet could help.
* Because NN + custom loss layer has loss output as a result, it is not clear, how to calculate other eval metrics during training.
* For the same reason, i had to keep two different neural networks, one for fit, another one for predict. May be there is better solution? Default output layers like SoftmaxOutput and others do not have this problem.
* Even if there is need to keep two different networks, it would be nice to abstract loss layer, for future reuse. I guess it should be bind/simple_bind somehow to base net, but wasn't able to how to use this syntax.
